### ISSUE TITLE
Merge extension method - Collection was modified; enumeration operation may not execute

### ISSUE THREAD

**Issue by rcollette** [tmenier/Flurl#314]
The problem is that the Where clause is using an enumerable in `for` loop where the enumerable is looking at the same collection that is being modified.   The quick fix for this is to append to the `Where()` statement a `ToArray()`
```
public static void Merge<TKey, TValue>(this IDictionary<TKey, TValue> d1, IDictionary<TKey, TValue> d2) {
            foreach (var kv in d2.Where(x => !d1.Keys.Contains(x.Key)))
                d1.Add(kv);
} 
```

```
Flurl.Http.FlurlHttpException: GET http://someurl failed. Collection was modified; enumeration operation may not execute. ---> System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
 at bool System.Collections.Generic.Dictionary<TKey, TValue>+Enumerator.MoveNext()
 at bool System.Linq.Enumerable+WhereEnumerableIterator<TSource>.MoveNext()
 at void Flurl.Util.CommonExtensions.Merge<TKey, TValue>(IDictionary<TKey, TValue> d1, IDictionary<TKey, TValue> d2)
 at void Flurl.Http.FlurlRequest.WriteHeaders(HttpRequestMessage request)
```

This is a pretty bad bug for us.

**Comment by tmenier** [comment]
Can you provide a code snippet that reproduces this?

**Comment by rcollette** [comment]
```
public async Task<string> RefreshToken(string authToken)
        {
            return (await _v1AdminClient
                    .Request("ping")
                    .WithHeader("authToken", authToken)
                    .GetAsync())
                .Headers.GetValues("authToken").First();
        }
```

**Comment by rcollette** [comment]
I'm also seeing this in calls that don't have a header specified in the request.

```
return await _v1AdminClient
                .Request("users")
                .AppendPathSegment(subscriptionId)
                .AppendPathSegment("idp")
                .AppendPathSegment(idpUserId)
                .GetJsonAsync<User>();
```

**Comment by rcollette** [comment]
I think I figured this out.  The `ToArray` mentioned above is a Red Herring.

The `WithBasicAuth`  used in our service constructor is applied each time the "cached" client is retrieved from the factory and that modifies the client's `Headers` dictionary.

When the Authorization header is added it is done using the index of the dictionary.

`.Headers["Authorization"]="the value"`

The original value is removed from the dictionary, the new value added and the internal version number of the dictionary is incremented.   Any in progress enumeration then checks the version number matches on the next `MoveNext` call and throws the exception if the `Headers` dictionary was modified since first getting the enumerator.

So basically, if we're making a request, and at the same time our service that sets the basic auth credentials is called, we get this conflict.

So the question becomes, how can the shared client be configured with default values, beyond the base path, in a manner that is thread safe?   

You have an example in #244  which shows setting a header in the service constructor, but I suspect that will have the same issues as here.

In our constructor we could check if the Authorization header already exists for the client before calling WithAuthorization, or just set it on each request.   Would be nice to have a "configure once" pattern that is simple to use though.

If my analysis is correct then the documentation (and maybe historical advice) for the current release might benefit from some guidance in this regard.   I don't know why but my first instinct was that I was configuring the client one time when using the factory and given the advice on 244 I'm probably not the only one to think that way.

Thanks for your time considering this issue.

**Comment by DarkChowder** [comment]
A possible solution is making headers a concurrent dictionary, which can be enumerated safely while being changed. The cookies dictionary in FlurlClient would also have this issue.

**Comment by tmenier** [comment]
Ah, I see, it's a concurrency problem. Will the credentials be the same for the entire life of the app, e.g. they're coming from some config source? If so you could configure the client in Startup rather than in your service ctor:

```c#
services.AddSingleton<IFlurlClientFactory>(_ => {
    var fac = new PerBaseUrlFlurlClientFactory();
    fac.Get(config.BaseUrl)
        .WithBasicAuth(config.Username, config.Password);
    return fac;
});
```

**Comment by tmenier** [comment]
@DarkChowder I might consider something like that, although I don't know if guaranteeing that FlurlClient is 100% threadsafe is necessarily something I'd consider important. It's generally accepted that unless you're dealing a type that was specifically designed for concurrency (like any concurrent collection), the consumer is on the hook for ensuring thread safety when modifying a singleton.

One idea I'm pondering that could give users a clearer path is exposing a [ConfigureClient](https://github.com/tmenier/Flurl/blob/master/src/Flurl.Http/FlurlHttp.cs#L37-L42) method on FlurlClientFactory, rather than just at the global static level as it is. This just simply puts a lock around a provided configuration lambda. This should prevent concurrency errors but still might not be ideal to call in the ctor of a transient service in a busy app. Still, it could make the example above nice and fluent:

```c#
services.AddSingleton<IFlurlClientFactory>(_ =>
    new PerBaseUrlFlurlClientFactory()
        .ConfigureClient(config.BaseUrl, cli =>
            cli.WithBasicAuth(config.Username, config.Password)));
```

**Comment by rcollette** [comment]
Funny I was thinking something along those lines this morning but was thinking about it in terms of passing a lambda to the client constructor (not sure that's even possible, never tried it), but the example you have above would be helpful in this specific case.

Couldn't `ConfigureClient` use a bool property of the client to determine if it was already configured (from itself) and then skip configuration?   Or just pass the lambda to the factory directly in which case it wouldn't call the lambda if an existing client was found (but OMG that wouldn't be fluent 😄 )

More important, I think is steering people away from using the fluent configuration directly on the client when using the factory in a non-singleton service.   One of those little yellow triangle warning like they use in MS documentation comes to mind.

I could be wrong but my impression is the `lock` isn't always the fastest mechanism. You might want to consider `SemaphoreSlim` or `ReaderWriterLockSlim` and generally a pattern something like:

If I need to update
    lock
         If I need to update
              Do the work

because if you are waiting on the lock and the prior lock holder did the update, then you don't want to do the update twice.

This article gets even deeper into `SemphoreSlim` lock performance, although in a brief 1 pass read I must admit I don't get what the `AsyncLock` class adds to the mix beyond what `SemphoreSlim` does by itself.
https://www.hanselman.com/blog/ComparingTwoTechniquesInNETAsynchronousCoordinationPrimitives.aspx

Thank you for already having pondered a solution a bit so quickly.   I don't mind you closing this thread since now we're talking about a potential feature/enhancement rather than a bug.

**Comment by DarkChowder** [comment]
The ConfigureClient on the factory solution could update the construction and then dump the cache without any locks by exchanging references on fields.

It would be good to return an immutable flurlclient interface from the factory that hides the mutable parts of the client, but it would be a breaking change.

**Comment by tmenier** [comment]
Thanks for all your thoughts. I need to ponder this a bit more. I know the locking pattern I chose is about as simple as it gets, but that was sort of the point. These are some good ideas and I'll consider them, but again, I'm more concerned with steering people away from configuring the same client from multiple threads and toward doing it once at startup.

**Comment by rcollette** [comment]
Dumping the cache would defeat the purpose of having a singleton HttpClient would it not?

**Comment by snekbaev** [comment]
Hi,

working on a small library. Wanted to hide Flurl and avoid explicit initialization call, ended up having a small factory wrapper:

```
 internal static class FlurlClientFactoryGetWrapper
    {
        private static readonly object SyncLock = new object();
        private static          bool   _isInitialized;

        public static IFlurlClient Get( IFlurlClientFactory flurlClientFactory, ISettings settings )
        {
            var client = flurlClientFactory.Get( settings.BaseUrl );

            PerformOneTimeClientConfiguration( client, settings );

            return client;
        }

        private static void PerformOneTimeClientConfiguration( IFlurlClient client, ISettings settings )
        {
            if (_isInitialized)
            {
                return;
            }

            lock (SyncLock)
            {
                if (_isInitialized)
                {
                    return;
                }

                if (settings.ConnectionLeaseTimeoutToRemote != null)
                {
                    client.Settings.ConnectionLeaseTimeout = settings.ConnectionLeaseTimeoutToRemote.Value;
                }

                client.WithHeader( "x-api-key", settings.ApiKey );
                client.WithHeader( "Accept", "application/json" );

                _isInitialized = true;
            }
        }
    }
```

Nothing special, but if somebody finds it useful :)

**Comment by tmenier** [comment]
@rcollette I'll take you up on closing this issue, I moved the parts I'm acting on to #356.
